File: claude-code-agents-typescript/src/models/improvement.ts
// src/models/improvement.ts — Improvement suggestion and selector patch models
import type { Confidence, SpiderName } from '../types.js';
import { toSpiderName } from '../types.js';
import type { PageType } from './crawl-target.js';

// ── Improvement Suggestion ──────────────────────────────────────
export interface ImprovementSuggestion {
  readonly spider: SpiderName;
  readonly selector: string;
  readonly issue: string;
  readonly proposedFix: string;
  readonly confidence: Confidence;
  readonly pageType: PageType | undefined;
  readonly impact: 'low' | 'medium' | 'high';
}

export function isHighConfidence(suggestion: ImprovementSuggestion): boolean {
  return (suggestion.confidence as number) >= 0.8;
}

export function toSelectorPatch(suggestion: ImprovementSuggestion): SelectorPatch {
  return createSelectorPatch({
    spider: suggestion.spider as string,
    oldSelector: suggestion.selector,
    newSelector: suggestion.proposedFix,
    rationale: suggestion.issue,
  });
}

// ── Selector Patch ──────────────────────────────────────────────
export interface SelectorPatch {
  readonly spider: SpiderName;
  readonly oldSelector: string;
  readonly newSelector: string;
  readonly rationale: string;
  readonly validated: boolean;
}

export function createSelectorPatch(input: {
  readonly spider: string;
  readonly oldSelector: string;
  readonly newSelector: string;
  readonly rationale?: string;
  readonly validated?: boolean;
}): SelectorPatch {
  return {
    spider: toSpiderName(input.spider),
    oldSelector: input.oldSelector,
    newSelector: input.newSelector,
    rationale: input.rationale ?? '',
    validated: input.validated ?? false,
  };
}

export function applyToSource(patch: SelectorPatch, source: string): string {
  if (!source.includes(patch.oldSelector)) {
    throw new Error(
      `Old selector '${patch.oldSelector}' not found in source for spider '${patch.spider}'`,
    );
  }
  return source.replace(patch.oldSelector, patch.newSelector);
}

export function asDiffLine(patch: SelectorPatch): string {
  return `- ${patch.oldSelector}\n+ ${patch.newSelector}`;
}

File: claude-code-agents-python/src/models/improvement.py
"""Improvement suggestion and selector patch models."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

__all__ = ["ImprovementSuggestion", "SelectorPatch"]


class ImprovementSuggestion(BaseModel):
    """A suggested improvement for a spider's extraction logic."""

    spider: str = Field(..., description="Name of the spider to improve")
    selector: str = Field(..., description="The CSS/XPath selector to change")
    issue: str = Field(..., description="Description of what is wrong")
    proposed_fix: str = Field(..., description="Proposed new selector or logic change")
    confidence: float = Field(
        default=0.5, ge=0.0, le=1.0, description="Confidence in this suggestion"
    )
    page_type: Optional[str] = Field(
        default=None, description="Page type this applies to"
    )
    impact: str = Field(
        default="medium",
        description="Expected impact level: low, medium, high",
    )

    @property
    def is_high_confidence(self) -> bool:
        """True if confidence >= 0.8."""
        return self.confidence >= 0.8

    def to_patch(self) -> SelectorPatch:
        """Convert to a SelectorPatch."""
        return SelectorPatch(
            spider=self.spider,
            old_selector=self.selector,
            new_selector=self.proposed_fix,
            rationale=self.issue,
        )


class SelectorPatch(BaseModel):
    """A concrete patch replacing one selector with another."""

    spider: str = Field(..., description="Name of the spider to patch")
    old_selector: str = Field(..., description="The current selector to replace")
    new_selector: str = Field(..., description="The replacement selector")
    rationale: str = Field(default="", description="Why this change is being made")
    validated: bool = Field(
        default=False, description="Whether this patch has been validated against live pages"
    )

    def apply_to_source(self, source: str) -> str:
        """Apply this patch to spider source code by replacing the old selector."""
        if self.old_selector not in source:
            raise ValueError(
                f"Old selector '{self.old_selector}' not found in source for spider '{self.spider}'"
            )
        return source.replace(self.old_selector, self.new_selector)

    def as_diff_line(self) -> str:
        """Format as a diff-style line for logging."""
        return f"- {self.old_selector}\n+ {self.new_selector}"
